web/js/stats/stats.js

```diff
@@ -0,0 +1,348 @@
+
+                const color = j > 0 && data[j] > data[j - 1] ? opts.style.barFallColor : opts.style.barColor;
+
+                drawRect(x, barHeight - Math.round(y), barWidth, barHeight, color);
+            }
+        }
+
+        const drawRect = (x, y, w, h, color = opts.style.barColor) => {
+            _context.fillStyle = color;
+            _context.fillRect(x, y, w, h);
+        }
+
+        return {add, get, max, render}
+    }
+
+    /**
+     * Get cached module UI.
+     *
+     * HTML:
+     * <div><div>LABEL</div><span>VALUE</span>[<span><canvas/><span>]</div>
+     *
+     * Returns exposed ui sub-tree and the _value as only changing node.
+     *
+     * @param label The name of the stat to show.
+     * @param withGraph True if to draw a graph.
+     * @param postfix The name of dimension of the stat.
+     * @returns {{el: HTMLDivElement, update: function}}
+     */
+    const moduleUi = (label = '', withGraph = false, postfix = 'ms') => {
+        const ui = document.createElement('div'),
+            _label = document.createElement('div'),
+            _value = document.createElement('span');
+        ui.append(_label, _value);
+
+        let postfix_ = postfix;
+
+        let _graph;
+        if (withGraph) {
+            const _container = document.createElement('span');
+            ui.append(_container);
+            _graph = graph(_container);
+        }
+
+        _label.innerHTML = label;
+
+        const withPostfix = (value) => postfix_ = value;
+
+        const update = (value) => {
+            if (_graph) _graph.add(value);
+            // 203 (333) ms
+            _value.textContent = `${value < 1 ? '<1' : value} ${_graph ? `(${_graph.max()}) ` : ''}${postfix_}`;
+        }
+
+        return {el: ui, update, withPostfix}
+    }
+
+    /**
+     * Latency stats submodule.
+     *
+     * Accumulates the simple rolling delta mean value
+     * between a server request and a following server response values.
+     *
+     *      window
+     *   _____________
+     *  |            |
+     * [1, 1, 3, 4, 1, 4, 3, 1, 2, 1, 1, 1, 2, ... n]
+     *              |
+     *    stats_snapshot_period
+     *    mean = round(next - mean / length % window)
+     *
+     * Events:
+     * <- PING_RESPONSE
+     * <- PING_REQUEST
+     *
+     * ?Interface:
+     *  HTMLElement get()
+     *  void enable()
+     *  void disable()
+     *  void render()
+     *
+     * @version 1
+     */
+    const latency = (() => {
+        let listeners = [];
+
+        let mean = 0;
+        let length = 0;
+        let previous = Date.now();
+        const window = 5;
+
+        // UI
+        const ui = moduleUi('Ping', true);
+
+        const onPingRequest = (data) => previous = data.time;
+
+        const onPingResponse = () => {
+            length++;
+            const delta = Date.now() - previous;
+            mean += Math.round((delta - mean) / length);
```

Comment:
oh first time know mean from stream can calculate this way :D . Intersting

Reply:
[Yeah](https://en.wikipedia.org/wiki/Moving_average).